cmd/cloudFoundryDeploy.go

```diff
@@ -668,21 +673,118 @@ func deployMta(config *cloudFoundryDeployOptions, mtarFilePath string, command c
 		cfDeployParams = append(cfDeployParams, deployParams...)
 	}
 
-	cfDeployParams = append(cfDeployParams, handleMtaExtensionDescriptors(config.MtaExtensionDescriptor)...)
+	extFileParams, extFiles := handleMtaExtensionDescriptors(config.MtaExtensionDescriptor)
 
-	return cfDeploy(config, cfDeployParams, nil, nil, command)
+	for _, extFile := range extFiles {
+		_, err := fileUtils.Copy(extFile, extFile+".original")
+		if err != nil {
+			return fmt.Errorf("Cannot prepare mta extension files: %w", err)
+		}
+		err = handleMtaExtensionCredentials(extFile, config.MtaExtensionCredentials)
+		if err != nil {
+			return fmt.Errorf("Cannot handle credentials inside mta extension files: %w", err)
+		}
+	}
+
+	cfDeployParams = append(cfDeployParams, extFileParams...)
+
+	err := cfDeploy(config, cfDeployParams, nil, nil, command)
+
+	for _, extFile := range extFiles {
+		renameError := fileUtils.FileRename(extFile+".original", extFile)
+		if err == nil && renameError != nil {
+			return renameError
+		}
+	}
+
+	return err
+}
+
+func handleMtaExtensionCredentials(extFile string, credentials map[string]interface{}) error {
+
+	log.Entry().Debugf("Inserting credentials into extension file '%s'", extFile)
+
+	b, err := fileUtils.FileRead(extFile)
+	if err != nil {
+		return errors.Wrapf(err, "Cannot handle credentials for mta extension file '%s'", extFile)
+	}
+	content := string(b)
+
+	env, err := toMap(_environ(), "=")
+	if err != nil {
+		errors.Wrap(err, "Cannot handle mta extension credentials.")
+	}
+
+	updated := false
+	missingCredentials := []string{}
+	for name, credentialKey := range credentials {
+		credKey, ok := credentialKey.(string)
+		if !ok {
+			return fmt.Errorf("Cannot handle mta extension credentials: Cannot cast '%v' (type %T) to string", credentialKey, credentialKey)
+		}
+		pattern := "<%= " + name + " %>"
+		if strings.Contains(content, pattern) {
+			cred := env[credKey]
```

Comment:
I think the lookup happens with the wrong key.
The env variable is named after the config key, not it's value (which is the ID of the secret and is the actual secret in the env-var). See also https://github.com/SAP/jenkins-library/pull/2432/files#diff-e35691724406d2191d9d957bef68787270bff32973e572651bdfe6e9bff9174cR238

```suggestion
			cred := env[name]
```

Reply:
We have in the config e.g.

```
cloudFoundryDeploy:
    org: myOrg
    space: s123
    mtaExtensionCredentials:
        xxxCredentialsHash: 'XXX_XXXXXXXX_XXX_XXX_HASH'
```

* `XXX_XXXXXXXX_XXX_XXX_HASH` is the id of the credentials entry as it is maintained in Jenkins. 
* `xxxCredentialsHash` is the id which needs to be provided inside the mta extension file (like `<%= xxxCredentialsHash %>`.

So outside Jenkins when working with environment variables we have to provide an environment variable with name `XXX_XXXXXXXX_XXX_XXX_HASH` and the secret text as the value. Lets say we have `export XXX_XXXXXXXX_XXX_XXX_HASH=abc123` then we will have inside the mta extension file `abc123` instead of `<%= xxxCredentialsHash %>`.

That all does not look wrong to me. When we use the `name` as suggested above we would have to expose an environment variable with the same name like the identifier in the mta ext file. With the approach as suggested above we don't have an indirection between the identifer in the mta ext file on the one hand and the name of the environment variable on the other hand. This would also mean the value (here: `XXX_XXXXXXXX_XXX_XXX_HASH`) is not needed anymore at all. It would have been possible to do it without indirection also in groovy. But the developer providing the code in groovy did it with indirection. Since this here is a shift from one language to another we should try to be close to what we had before.

I believe there is some kick-up in the way we do it right now in the groovy layer when preparing to expose the environment variables.

Corresponding log entries when I run the code with an example project on my end is:

```
debug cloudFoundryDeploy - Inserting credentials into extension file 'myExt'
debug cloudFoundryDeploy - Mta extension credentials handling: Placeholder 'xxxCredentialsHash' has been replaced by credential denoted by 'XXX_XXXXXXXX_XXX_XXX_HASH' in file 'myExt'
debug cloudFoundryDeploy - Mta extension credentials handling: Extension file 'myExt' has been updated.
warn  cloudFoundryDeploy - mta extension credential handling: Unresolved placeholders found after inserting credentials: [<%= xxxCredentialsHashUnkown %>]
```
Correct me if I'm wrong ...

Reply:
> * `XXX_XXXXXXXX_XXX_XXX_HASH` is the id of the credentials entry as it is maintained in Jenkins.
> * `xxxCredentialsHash` is the id which needs to be provided inside the mta extension file (like `<%= xxxCredentialsHash %>`.

Yes.

> So outside Jenkins when working with environment variables we have to provide an environment variable with name `XXX_XXXXXXXX_XXX_XXX_HASH` and the secret text as the value. Lets say we have `export XXX_XXXXXXXX_XXX_XXX_HASH=abc123` then we will have inside the mta extension file `abc123` instead of `<%= xxxCredentialsHash %>`.

I think you are wrong here. I have snipped the rest of your reply, because I find it much too complicated.

Let me repeat the situation in my own terms, which hopefully make it obvious what is needed where:

---- config ---

```yaml
cloudFoundryDeploy:
    mtaExtensionCredentials:
        userProvidedKey1: 'jenkinsCredentialStoreId1'
        userProvidedKey2: 'jenkinsCredentialStoreId2'
```

---- mta extension file ---

```
...
<%= userProvidedKey1 %>
...
````

The requirement is that we replace the placeholder "<%= userProvidedKey1 %>" in the mta credentials file with the actual secret, which we obtain from the Jenkins secret store via the value of that key in the user config.

So job of Jenkins wrapper is: Replace "jenkinsCredentialStoreId1" and "jenkinsCredentialStoreId2" with actual secrets and expose as env variables:

userProvidedKey1=actualSecret1
userProvidedKey2=actualSecret2

This is also what your Jenkins wrapper PR currently does. I have linked the code line where it happens.

The job of the Go part is: Find the env variables named userProvidedKey1 and userProvidedKey2, and insert their values in the place-holders. The configured jenkinsCredentialStoreIds are not relevant at this point anymore. There were only relevant in the Jenkins wrapper to resolve the actual secrets.

The error message from the test IMHO fits directly to my explanation and the code where I think it is wrong.

Also important: The user provided keys are guaranteed to be unique (as they are map keys), while two or more keys may reference the same secret store ID. That is also a reason why I think my interpretation is correct.

Reply:
> This is also what your Jenkins wrapper PR currently does. I have linked the code line where it happens.

Saw that you changed it to expose the variable under the secretID instead of the user provided key instead. That will fix the problem as well, of course. I am not sure what I find more elegant. If the go code somehow resolves the secret IDs itself, then it needs to process the IDs. But of course then it doesn't need the env variables in the first place.

Reply:
>The job of the Go part is: Find the env variables named userProvidedKey1 and userProvidedKey2, and insert their values in the place-holders. The configured jenkinsCredentialStoreIds are not relevant at this point anymore. There were only relevant in the Jenkins wrapper to resolve the actual secrets.

This is something else that what is done in the groovy layer. In the groovy layer we have the indirection. Of course we have the freedom to get ride of that indirection, but when shifting from one language to another it is a good practive to mimic the behaviour.

We have the config

```
placeholderNameWhichGetReplacedInMtaExt: nameOfCredentialKey.
```

And the credential key now outside groovy is in my understanding the `nameOfCredentialKey`.


Your approach and my approach will work. For your case need to adjust in the go part and for my approach we have to adjust in the groovy part. My approach is IMO closer to what we have in groovy. With your approach the values are not used anymore.

I don't know for sure, but I can imagine having that indirection can be a benefit. Why do we have that indirection in groovy. Also here is would be possible to assume that the credentials are maintained with a key matching to the variable name as it gets replaced in the mta ext file.

Reply:
I am fine with your solution. I will apply it to my other branch and run another test.